### DISCUSSION TITLE
[QUESTION] Document indices extended in add_document function in IndexedDatasetBuilder but not in add_item

### DISCUSSION THREAD

**Discussion by clarence-lee-sheng** [NVIDIA/Megatron-LM#1073] [Q&A]
Dear Megatron-LM team 

In the IndexedDatasetBuilder class add_item function (https://github.com/NVIDIA/Megatron-LM/blob/912be7889298da7b3dbf39dea5b4441c1e412488/megatron/core/datasets/indexed_dataset.py#L767-L779), the document indices is not extended, however for the add_document function, the document indices are extended (https://github.com/NVIDIA/Megatron-LM/blob/912be7889298da7b3dbf39dea5b4441c1e412488/megatron/core/datasets/indexed_dataset.py#L781-L798). May I ask why is that the case and when do we use add_item? 

Thank you!

**Comment by Advait251206** [comment]
The distinction comes down to what each method is intended to represent.

- `add_item()` appends a **single sequence/item** to the dataset.
- `add_document()` appends an **entire document**, which may consist of one or more sequences, and records where that document ends.

The `document_indices` array is used to track **document boundaries**, not individual sequences. That's why `add_document()` updates it:

```python
self.document_indices.append(len(self.sequence_lengths))
```

This records the cumulative number of sequences after the document has been added, allowing the dataset loader to reconstruct document boundaries later.

In contrast, `add_item()` only updates:

- the raw data,
- `sequence_lengths`,
- and optionally `sequence_modes`.

It intentionally does **not** update `document_indices`, because a single item isn't necessarily intended to represent a complete document. Multiple calls to `add_item()` could conceptually belong to the same document, so automatically extending `document_indices` after every item would effectively treat every item as a separate document, which isn't always correct.

As for when `add_item()` is used, it's generally intended for incremental dataset construction where items are written one at a time. If your downstream processing depends on document boundaries (for example, document-level sampling or preserving document context), you'll typically want to use `add_document()` instead so that `document_indices` is populated correctly.

So the differing behavior appears to be intentional: `document_indices` tracks **documents**, while `sequence_lengths` tracks **individual sequences**.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
